Implement createSettingsStore helper for reusable settings stores

felixarntz opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #1249

## Relevant technical choices

* This PR not only allows creating a store with the generic settings functionality, it also automatically includes the specific `set*( value )` actions and `get*()` selectors for sub-settings.
* This will significantly reduce the amount of work necessary to implement stores like in #1224 and #1247, and furthermore ensure full consistency.
* Adds optional `options` parameters to the existing `createModuleStore` and `createNotificationsStore` for additional optional arguments.
* Includes determining the store name (unless provided via `options`) in the creator functions.
* Includes awareness of the registry (defaults to `googlesitekit-data` registry) in order to allow for things like using `select` within an action.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

felixarntz commented:
@tofumatt We need to use a different approach here to make the individual `get*` selectors fire the resolver for `getSettings` somehow (we can't `Data.select( STORE_NAME ).getSettings()` here since `STORE_NAME` is unknown at this point). Let's see how we can accomplish this. I put some code into the PR so far, which _may_ work, but I'm not sure :)

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.js:
```diff
@@ -0,0 +1,392 @@
+		 */
+		*saveSettings() {
+			try {
+				const savedValues = yield actions.fetchSaveSettings();
```
```suggestion
To get store values inside the action you'll need to access the registry, but this is okay 🙂:

```js
				// TODO: We'll need access to the current store name in this code, which we don't right now.
				const updatedSettings = Data.select( STORE_NAME ).getSettings();
				const savedValues = yield actions.fetchSaveSettings( updatedSettings );
```

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.js:
```diff
@@ -0,0 +1,392 @@
+		[ FETCH_SETTINGS ]: () => {
+			return API.get( type, identifier, datapoint );
+		},
+		// TODO: How do we get the current values to be available here?
```
Arguments passed to actions are also passed to their relevant controls: https://developer.wordpress.org/block-editor/packages/packages-data/#registering-a-store has an example of this.

So any settings to be `POST`ed to the API should be provided to the action creator and will be available to the control. I left a comment about for how that would work 🙂

felixarntz replied:
Implemented this, alongside some other changes to make that possible (`storeName` and `registry` awareness of these functions).

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.js:
```diff
@@ -0,0 +1,392 @@
+					},
+					settings: {
+						...values,
+						// In case settings were already changed, they should take precedence.
```
Nice 👍

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.js:
```diff
@@ -0,0 +1,392 @@
+			}
+
+			default: {
+				// Check if this action is for a sub-setting reducer.
```
I'm not exactly sure how this code is intended to work, but I think it'll make more sense when there are some tests around it. I don't quite follow what this means yet, but that's okay 😄

felixarntz replied:
The code further down iterates through all the passed setting slugs (e.g. `accountID`, `propertyID` and so on in the case of Analytics module), and then automatically creates actions like `setAccountID` (with type `SET_ACCOUNT_ID`) and selectors like `getAccountID`. In order for those to work it also needs to modify the overall reducer via this `settingReducers` variable.

So here in the `default` case, the main reducer checks whether there's a reducer defined for the current action type (e.g. `SET_ACCOUNT_ID`).

felixarntz commented:
@tofumatt I think this should actually be in a working state. Would be great to get another review, then I'll work on tests to ensure it's actually in a working state.

tofumatt commented on assets/js/googlesitekit/data/create-notifications-store.js:
```diff
@@ -40,12 +40,24 @@ const RECEIVE_NOTIFICATIONS_FAILED = 'RECEIVE_NOTIFICATIONS_FAILED';
+ * @param {string} identifier        The data identifier, eg. a module slug like 'search-console'.
+ * @param {string} datapoint         The endpoint to request data from, e.g. 'notifications'.
+ * @param {Object} options           Optional. Options to consider for the store.
+ * @param {number} options.storeName Store name to use. Default is '{type}/{identifier}'.
```
Nice! I like the sane default here 👍

tofumatt commented on assets/js/googlesitekit/data/utils.js:
```diff
@@ -187,6 +187,25 @@ export const collectSelectors = collect;
+ * @param {...string} args A list of store names, all of which must be equal.
+ * @return {string} The single store name.
+ */
+export const collectName = ( ...args ) => {
```
This doesn't so much collect the names as verify them. It doesn't seem to be collecting any data–just verifying.

Maybe we want to change it to `checkStoreNames` or `ensureStoreNamesMatch`?

felixarntz replied:
I thought about that, but in a way it still collects the names and gives you the end result. Technically it's indeed more verifying than collecting, but I think this is more intuitive in combination with the other functions. As a developer, you basically just have to remember "for every piece of the store I need to use `collect{typeOfPiece}`. So I'd prefer to keep this name.

If it was a more standalone-intended utility function, I would agree that it should be something more about verifying.

tofumatt replied:
Fair enough, that point about not having to remember a different name for this is a good one. 👍

tofumatt commented on assets/js/googlesitekit/data/index.js:
```diff
@@ -51,6 +52,6 @@ Data.collectReducers = collectReducers;
 Data.collectResolvers = collectResolvers;
 Data.collectSelectors = collectSelectors;
 Data.collectState = collectState;
-Data.collectReducers = collectReducers;
```
Noting here for future reviewers that this was removed because it's a repeat assignment—it's already on line 52.

tofumatt commented:
(Seems sometimes the tests don't run coverage in watch mode, which I think I know the cause of, but for now you'll definitely get test coverage from `npm run test:js` and then opening the coverage report using `open coverage/index.html`. I'll try to tidy that up later, but for now that should do the trick 😅 )

felixarntz commented:
@tofumatt This is ready for full review now, I've added test coverage for the new functionality. It's probably easiest to review the individual commits, specifically regarding the few things I fixed as a result of the tests uncovering them as broken.

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+		describe( 'fetchSettings', () => {
+			it( 'does not require any params', () => {
+				expect( () => {
+					dispatch.fetchSettings();
```
I'm surprised this doesn't cause the console to error, given it should cause a fetch request to fire. 🤷‍♂

felixarntz replied:
It actually seems not to - could this be because I call `fetchSettings` directly here (the actual call is in the control)? Not too sure about the `wp.data` internals regarding this...

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+				// Two fetch requests, one to get settings, the other to update.
+				expect( fetch ).toHaveBeenCalledTimes( 2 );
+
+				expect( store.getState().settings ).toMatchObject( response );
```
Interesting. I consider this unexpected behaviour. The client (should) fetch settings before setting them, but what if you present the user an interface where the can change parts of the data on the client without server involvement? I suppose I'm struggling to find an example other than a list where they could add items while existing ones were fetched, but it's a bit surprising this happens.

That said: it's codified and thus expected, so that's good. And I suppose we likely want loading screens that prevent user interaction until they can save settings, so fair enough 👍

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+		describe( 'setSetting', () => {
+			it( 'has the correct action name', () => {
+				expect( Object.keys( storeDefinition.actions ) ).toEqual(
+					expect.arrayContaining( [ SETTING.action ] )
```
I know this variable was defined above, but I think this abstraction makes it a lot harder to understand precisely what this is testing. Consider:

```suggestion
					expect.arrayContaining( [ 'setTestSetting' ] )
```

I know it is "redundant" in the strictest sense, but the test as-is compares two variables not in-view in my VSCode window on a 5K display, so I find it to lack context compared to my suggestion.

tofumatt replied:
(The comment helps a bit, but I think we can be clearer 😄)

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+				expect( select.getSettings() ).toEqual( settings );
+			} );
+
+			it( 'returns client settings even if server settings have not loaded', () => {
```
If settings from the server simply haven't been loaded yet (and we know that… and we overwrite client settings after fetching from the server)… why not just have this also return `undefined`? It seems odd to let the client save settings we'll never use.

felixarntz replied:
We don't actually override client side settings when they're loaded from the server (see `receiveSettings`). We only override client side settings when they were just saved (see `receiveSaveSettings`).

Reasoning: When you save, the UI should be blocked, so we should safely assume server-side settings are in sync with client-side. When the server-side settings are being loaded though, depending on the implementation I'd say it may be acceptable here to already allow for user input.

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+		describe( 'getSetting', () => {
+			it( 'has the correct selector name', () => {
+				expect( Object.keys( storeDefinition.selectors ) ).toEqual(
+					expect.arrayContaining( [ SETTING.selector ] )
```
Similarly, I'd find this a bit more informative with the expected contents inline in the test:

```suggestion
					expect.arrayContaining( ['getTestSetting' ] )
```

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+import { createSettingsStore } from './create-settings-store';
+
+const SETTING = {
+	slug: 'testSetting',
```
Rather than calling this `testSetting`, using a more illustrative example might help with readability and understanding of how this works. The word "setting" is already a bit vague and is used a lot throughout this code.

Even something not-serious like `isSkyBlue` or something might be a bit more illustrative. Just a thought! 😄

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+				);
+
+				expect( fetch ).toHaveBeenCalledTimes( 1 );
+				expect( select[ SETTING.selector ]() ).toEqual( value );
```
Again: I know this is correct and semantically the same as `select.getTestSetting()`, but `select.getTestSetting()` would be a lot more readable and would better illustrate what's happening here.
```suggestion
				expect( select.getTestSetting() ).toEqual( value );
```

tofumatt commented on assets/js/googlesitekit/data/create-settings-store.test.js:
```diff
@@ -0,0 +1,453 @@
+				fetch
+					.mockResponseOnce( async ( req ) => {
+						if ( req.url.startsWith( `/google-site-kit/v1/${ type }/${ identifier }/data/${ datapoint }` ) ) {
+							return Promise.resolve( {
```
Since this is an async function already, I think you should be able to just return the object here without the `Promise.resolve()` bit. Not a biggie, but might clean things up a bit 🤷‍♀

felixarntz commented:
@tofumatt Ready for re-review!
